homeassistant/components/firmata/binary_sensor.py

```diff
@@ -0,0 +1,84 @@
+"""Support for Firmata binary sensor input."""
+
+import logging
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.const import CONF_NAME
+
+from .const import CONF_NEGATE_STATE, CONF_PIN, CONF_PIN_MODE, DOMAIN
+from .pin import FirmataBinaryDigitalInput, FirmataPinUsedException
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Firmata binary sensors."""
+    new_entities = []
+
+    boards = hass.data[DOMAIN]
+    board = boards[config_entry.entry_id]
+    for binary_sensor in board.binary_sensors:
+        try:
+            binary_sensor_entity = FirmataBinarySensor(
+                board, config_entry, **binary_sensor
+            )
+            new_entities.append(binary_sensor_entity)
+        except FirmataPinUsedException:
+            _LOGGER.error(
+                "Could not setup binary sensor on pin %s since pin already in use.",
+                binary_sensor[CONF_PIN],
+            )
+    async_add_entities(new_entities)
+
+
+class FirmataBinarySensor(BinarySensorEntity):
+    """Representation of a binary sensor on a Firmata board."""
+
+    def __init__(self, board, config_entry, **kwargs):
+        """Initialize the binary sensor."""
+        self._name = kwargs[CONF_NAME]
+        self._state = None
+        self._config_entry = config_entry
+        self._conf = kwargs
+
+        pin = self._conf[CONF_PIN]
+        pin_mode = self._conf[CONF_PIN_MODE]
+        negate = self._conf[CONF_NEGATE_STATE]
+
+        self._location = (DOMAIN, self._config_entry.entry_id, "pin", pin)
+        self._unique_id = "_".join(str(i) for i in self._location)
+
+        self._api = FirmataBinaryDigitalInput(board, pin, pin_mode, negate)
+
+    async def async_added_to_hass(self):
+        """Set up a binary sensor."""
+        await self._api.start_pin(self.async_write_ha_state)
+
+    async def async_will_remove_from_hass(self):
+        """Stop reporting a binary sensor."""
+        await self._api.stop_pin()
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if binary sensor is on."""
+        return self._api.is_on
+
+    @property
+    def name(self) -> str:
+        """Get the name of the pin."""
+        return self._name
+
+    @property
+    def should_poll(self) -> bool:
+        """No polling needed."""
+        return False
+
+    @property
+    def unique_id(self):
+        """Return a unique identifier for this device."""
+        return self._unique_id
+
+    @property
+    def device_info(self) -> dict:
+        """Return device info."""
+        return self._api.board_info
```

Comment:
The `device_info` is an entity api and we should define the info in the entity class, not in the board api. It's ok to set attributes in the board api with the used values, but we need to define the interface here in the entity.

Reply:
The board also uses the device registry using `device_registry.async_get_or_create()`. This way it still shows up as a device it is configured without any entities. Since the entities and the device must share the exact same device attributes, it seemed cleaner to define it only in one place. Where/how should I define it? I put it in the board API because it is dependent on and specific to the board.

Reply:
Why do we want to create a device when there's no entities? I think we can simplify and remove that.

Reply:
Where should this check be added? It would I guess need to be checked in both the config flow and `async_setup` so that neither new devices or updated devices are created.

Reply:
If we just implement `device_info` in the entity class, there's no check needed.

Reply:
I don't think we want to connect to the device or setup at all if it will not appear in the UI and no entities are created. This would make it more difficult for a user to debug a connectivity issue if they are simply trying to get it to connect.

Reply:
I'm not following. What part of the code are we talking about now?

Reply:
#### Currently as it is implemented:

User could config something like this:
```yaml
firmata:
  - serial_port: /dev/serial/by-id/usb-Teensyduino_USB_Serial_358320-if00
    serial_baud_rate: 57600
```
This causes the device to appear in the integrations page but without any entities, as they haven't added any binary sensors or switches. However, a connection to the device is still made and setup and appears in the UI. The device is shown from the `device_registry.async_get_or_create()` call made from the board. Clicking "remove" in the UI shuts down the device as expected.

This allows a user to get the connection working before they add any entities. I think this is the same as at least some other similar integrations. One example: you can add an esphome device even if it doesn't have any entities. Not trying to argue what is right (that's your decision lol), just an example.

If the user configs like this:
```yaml
firmata:
  - serial_port: /dev/serial/by-id/usb-Teensyduino_USB_Serial_358320-if00
    serial_baud_rate: 57600
    switches:
      - name: my_light
        pin_mode: OUTPUT
        pin: 4
        negate: true
```
Then the device appears both from `device_registry.async_get_or_create()` and the `my_light` entity as expected.


#### If `device_registry.async_get_or_create()` is removed:

User could config something like this:
```yaml
firmata:
  - serial_port: /dev/serial/by-id/usb-Teensyduino_USB_Serial_358320-if00
    serial_baud_rate: 57600
```
The device then *does not* appear in the UI since there are no entities, but the board is still connected to and setup. This is deceiving to the user and looks like a connection issue since nothing appears in the UI but the the board is in fact connected. Additionally, nothing appears in the logs since it connected properly.

I see your point about not connecting to the board if there are no entities though if you would like to go that route. I could trigger a warning to the logs in that case. So, should that check be implemented in the voluptuous config check (seems tricky/messy)? Or should it be implemented in `async_setup` and `async_step_import` so that the entity isn't created in the first place and so that it still isn't connected to if it is updated later?

Reply:
Though I am also a bit confused by why the `device_info` for the entity can't rely on the board if it is literally grouped with the board and what identifies it is the board, not the entity itself. Here is how it currently appears in HA, which is the same as other similar integrations I am using:

![image](https://user-images.githubusercontent.com/4472083/87183689-95b83280-c2b4-11ea-9de9-1765064eaf4d.png)
![image](https://user-images.githubusercontent.com/4472083/87183797-d0ba6600-c2b4-11ea-8f3a-45190e925c51.png)

Reply:
Besides the device_info, the board doesn't use any core home assistant APIs. If we make the separation so that the board class isn't aware of any home assistant APIs we have a good separation of concerns and less entangling. That's why I want to remove the device_info API awareness from the board class.

Regarding allowing users to set up the integration without entities, I don't see the point to set up the integration without entities. But if we still want to do that, we can create the home assistant device in `async_setup_entry` in the base integration module after setting up the board. We don't need to mix in the home assistant API in the board class.

Reply:
> Besides the device_info, the board doesn't use any core home assistant APIs. If we make the separation so that the board class isn't aware of any home assistant APIs we have a good separation of concerns and less entangling. That's why I want to remove the device_info API awareness from the board class.

I see, makes sense.

> Regarding allowing users to set up the integration without entities, I don't see the point to set up the integration without entities. But if we still want to do that, we can create the home assistant device in `async_setup_entry` in the base integration module after setting up the board. We don't need to mix in the home assistant API in the board class.

OK, I think it is simplest and most intuitive for the user if I move it there.

But now back to what this conversation was originally about: where should the device attributes be stored/shared?

Reply:
It's fine to store the device info values as attributes of the board instance, eg the firmware and model etc.